File: Algorithms/Searching/binary-search.py
import os
import random

# Binary Search Algorithm
def binary_search(collection, target):

    beg = 0
    end = len(collection)-1
    mid = (beg + end) // 2

    # Iterate thorough all items in the collection
    while beg <= end:

        if collection[mid] == target:
            return mid

        elif collection[mid] > target:
            end = mid - 1

        elif collection[mid] < target:
            beg = mid + 1 

        mid = (beg + end) // 2
        
    # Target not found
    return None


if __name__ == "__main__":
    GREEN = '\033[92m'
    RESET = '\033[0m'
    print(f"{GREEN}Binary Search{RESET}")
    
    # Create random collection of items
    data = [random.randint(1, 40) for i in range(10)]

    # Sort the data
    data.sort()

    print(data)

    # Get the target
    target = int(input("Enter data to find : "))

    # Call binary search
    result = binary_search(data, target)

    # Check if data was found
    if result is None:
        print("Data not found")
    else:
        print(f"Found at index : {result}")

File: Algorithms/Searching/binary-search.rs
fn binary_search(nums: &[i32], target: i32) -> Option<usize> {
    let (mut beg, mut end): (usize, usize) = (0, nums.len() - 1);

    while beg <= end {
        let mid: usize = (beg + end) / 2;

        if nums[mid] == target {
            return Some(mid)
        }

        if nums[mid] > target {
            end = mid - 1;
        } else {
            beg = mid + 1
        }
    }

    return None
}

fn main() {
    let nums: [i32; 10] = [63, 83, 16, 10, 74, 66, 17, 1, 0, 60];

    let target = 63;
    match binary_search(&nums, target) {
        Some(num) => println!("{target} found at {num}"),
        None => println!("Not found"),
    }
}
